### ISSUE TITLE
Self-hosted Gitlab gives error: repository "may be private"

### ISSUE THREAD

**Issue by JamesCullum** [gitpod-io/gitpod#2686]
### Describe the bug
On my self-hosted Gitlab instance, I click the gitpod button and am redirected to https://gitpod.io/#https://gitlab.csirt.panasonic.eu/cybersecurity/docker-host/incident-response/-/tree/master/ (removing the parts for the tree does not make a difference)

This produces the following error:
```
<repo-name> may be private

The repository <repo-name> could not be accessed though you allowed to access private repositories.

You can try this ...
    Make sure the repository <repo-name> exists.
    If <repo-name> belongs to an organization, check if Gitpod was approved.
```

The link to "check if Gitpod was approved" leads to the OAuth application page, which does not reference an organization. An organization in Gitlab does not have a way to authorize an Oauth application from what I can see. The repository exists and the link to it works.

### Steps to reproduce
Try to create workspace with above mentioned repository, consisting of organization, group and repository name.

Gitlab version is 13.5.1 CE

### Expected behavior
Workspace should be accessible, given that the application is created for an admin user of both Gitlab and the organization with full access to the repoisitory.

### Example repository
https://gitlab.csirt.panasonic.eu/cybersecurity/docker-host/incident-response/-/tree/master/

**Comment by csweichel** [comment]
The installation is not available from Gitpod. When I try to access the repository link you provided, I get a 403 error.

Gitpod needs to be able to access the Gitlab installation.

**Comment by JamesCullum** [comment]
Hey @csweichel,

Gitpod is whitelisted via its IP ranges as below.
This worked so far to establish the connection, on the webserver I don't see a failed connection.
Should I whitelist your IP as well for testing?

```
<If "-R '34.0.0.0/8' || -R '35.0.0.0/8'">
	<If "%{HTTP_USER_AGENT} == 'gitpod.io' || %{HTTP_USER_AGENT} == 'gitbeaker'">
		SetEnvIf _ .* server_access=true
		SetEnvIf _ .* internal_access=gitpod
	</If>
</If>
```

**Comment by AlexTugarev** [comment]
@JamesCullum, the authenticator part of Gitpod uses that UA indeed, that's why login is working.

I'm afraid, other parts like the context parser do not send this header. The context parser throws the error from ☝🏻

**Comment by JamesCullum** [comment]
Hey @AlexTugarev, the webserver logs do not show any requests that are blocked when I attempt a log in. On the other side, it shows a redirect from Gitlab to the sign-in page.

See webserver log below
```
Start Time,Payload
"Jan 7, 2021, 11:06:53 AM","gitlab.csirt.panasonic.eu 34.78.117.67 - - [07/Jan/2021:10:06:01 +0000] ""GET /users/sign_in HTTP/1.1"" 200 443 9566 ""-"" ""gitbeaker"""
"Jan 7, 2021, 11:06:53 AM","gitlab.csirt.panasonic.eu 34.78.117.67 - - [07/Jan/2021:10:06:00 +0000] ""GET /api/v4/projects/cybersecurity%2Fdocker-host%2Fincident-response%2Ftree%2Fmaster HTTP/1.1"" 302 443 113 ""-"" ""gitbeaker"""
"Jan 7, 2021, 11:06:53 AM","gitlab.csirt.panasonic.eu 34.78.117.67 - - [07/Jan/2021:10:06:00 +0000] ""GET /users/sign_in HTTP/1.1"" 200 443 9566 ""-"" ""gitbeaker"""
"Jan 7, 2021, 11:06:53 AM","gitlab.csirt.panasonic.eu 34.78.117.67 - - [07/Jan/2021:10:06:00 +0000] ""GET /api/v4/projects/cybersecurity%2Fdocker-host%2Fincident-response%2Ftree HTTP/1.1"" 302 443 113 ""-"" ""gitbeaker"""
"Jan 7, 2021, 11:06:53 AM","gitlab.csirt.panasonic.eu 34.78.117.67 - - [07/Jan/2021:10:06:00 +0000] ""GET /users/sign_in HTTP/1.1"" 200 443 9566 ""-"" ""gitbeaker"""
"Jan 7, 2021, 11:06:53 AM","gitlab.csirt.panasonic.eu 34.78.117.67 - - [07/Jan/2021:10:06:00 +0000] ""GET /api/v4/projects/cybersecurity%2Fdocker-host%2Fincident-response HTTP/1.1"" 302 443 113 ""-"" ""gitbeaker"""
"Jan 7, 2021, 11:06:53 AM","gitlab.csirt.panasonic.eu 34.78.117.67 - - [07/Jan/2021:10:05:59 +0000] ""GET /users/sign_in HTTP/1.1"" 200 443 9566 ""-"" ""gitbeaker"""
"Jan 7, 2021, 11:06:53 AM","gitlab.csirt.panasonic.eu 34.78.117.67 - - [07/Jan/2021:10:05:59 +0000] ""GET /api/v4/projects/cybersecurity%2Fdocker-host HTTP/1.1"" 302 443 113 ""-"" ""gitbeaker"""
"Jan 7, 2021, 11:06:53 AM","gitlab.csirt.panasonic.eu 35.205.4.73 - - [07/Jan/2021:10:05:58 +0000] ""GET /api/v4/user HTTP/1.1"" 200 443 883 ""-"" ""gitbeaker"""
"Jan 7, 2021, 11:06:53 AM","gitlab.csirt.panasonic.eu 35.205.4.73 - - [07/Jan/2021:10:05:58 +0000] ""POST /oauth/token HTTP/1.1"" 200 443 252 ""-"" ""gitpod.io"""
"Jan 7, 2021, 11:06:53 AM","gitlab.csirt.panasonic.eu 35.240.23.89 - - [07/Jan/2021:10:05:57 +0000] ""GET /api/v4/user HTTP/1.1"" 200 443 883 ""-"" ""gitbeaker"""
"Jan 7, 2021, 11:06:53 AM","gitlab.csirt.panasonic.eu 35.240.23.89 - - [07/Jan/2021:10:05:57 +0000] ""POST /oauth/token HTTP/1.1"" 200 443 236 ""-"" ""gitpod.io"""
```

**Comment by AlexTugarev** [comment]
> the webserver logs do not show any requests that are blocked when I attempt a log in. 

Correct. Because the UA header is sent.

> I will add other UAs as I encounter them.

I don't expect any other UA header to be sent, unfortunately. That's the problem in you case. Once you are logged in and try to start a workspace, the context parser will try to do Gitlab API requests to retrieve information from the repo.

**Comment by JamesCullum** [comment]
Hey Alex, I'm not sure I understand what the issue is. All requests from Gitpod pass through, but the Gitlab rejects the token. Do you mean that this is related to the whitelisting or not?

**Comment by AlexTugarev** [comment]
I'm happy to rephrase https://github.com/gitpod-io/gitpod/issues/2686#issuecomment-756029652, just to clarify: `<repo-name> may be private` won't go await as long as the `server` component of Gitpod is behind the firewall mentioned above.

It's great that you figured how to make the login working behind the FW. You've created a rule to allow requests with a specific request header. This request header was configured to be sent, because some OAuth2 servers (e.g. from github.com) require them, or they might reject those clients.

Okay, now after a successful login, if you want to start a workspace on a github.com repository, the server needs to fetch some data using the GitHub API to compute the context properly. These API requests will fail, as they won't include the request header to be allowed in the FW. Given that the "read repository" request fails using a properly obtained token, the assumed condition is, that it's privat.

**Comment by JamesCullum** [comment]
Hey @AlexTugarev, I'm sadly still not sure if I completely get the issue. You mention github.com a few times, but we are both talking about Gitlab, right?

You mention that `These API requests will fail, as they won't include the request header to be allowed in the FW.`. It sounds like they fail because they do not have these headers. As mentioned, the FW does not block any such request and the log doesn't show anything like this.

I've now set the Gitlab to be publicly available and still get the same error. The repository is within a group and set to private - could this be the reason? (even though it says that I've granted access to private repositories)
